### PR TITLE
Problem (Fix #82): reward mechanisms not defined precisely

### LINKED ISSUE #82: Problem: reward mechanisms not defined precisely
need fairness explanation / update how votes are processed
--
define precisely (somewhat "formally") in relation to Tendermint ABCI what each part does, so that this can be easily checked against the implementation:
what are data inputs from where/when
how are they processed
what are outputs and when they are emitted

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- Add reward mechanisms in terms of abci events and rust code

### CODE REVIEW

**Review by devashishdxt — CHANGES_REQUESTED** [review]
This documentation should be detailed enough for non-developers to understand how rewards work. I guess, it'd be good to provide an overview of everything in "English" (for non-developers) and also in "Pseudo-code" (for developers).

Currently, it just copy-pastes the implementation which is as good as reading code.

**Comment by yihuang** [comment]
> This documentation should be detailed enough for non-developers to understand how rewards work. I guess, it'd be good to provide an overview of everything in "English" (for non-developers) and also in "Pseudo-code" (for developers).
> 
> Currently, it just copy-pastes the implementation which is as good as reading code.

The abstract part should be enough for non-developers?
For the implementation part, I find the rewarding implementation is quite straight forward, the real code is already readable enough for developers I think.

File: docs/modules/rewarding.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
+# Rewarding
+
+## Abstract
+
+Rewarding module enable crypto chain to reward protocol participators, the design should meet following requirements:
+
+- Fixed total supply.
+- Fairness, all honest participators should get rewards, non-participators should not.
+- The rewards amount a validator received should be proportional to the voting power.
```
Comment (tomtau) [inline docs/modules/rewarding.md]: the last point is a bit unclear -- the emitted amount to all validators is proportional to the total vote power. The received share should be proportional only to participation?

File: docs/modules/rewarding.md [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+TODO
+
+## Implementation
+
+### Network parameters
+
+```rust
+pub struct RewardsParameters {
+    /// Time inteval in seconds to do rewards distribution
```
Comment (tomtau) [inline docs/modules/rewarding.md]:
```suggestion
    /// Time interval in seconds to do rewards distribution
```

File: docs/modules/rewarding.md [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    pub reward_period_seconds: u64,
+
+    /// Maximum monetary expansion for rewards.
+    pub monetary_expansion_cap: Coin,
+    /// Monetary expansion formula parameter
+    pub monetary_expansion_r0: Milli,
+    /// Monetary expansion formula parameter
+    pub monetary_expansion_tau: u64,
+    /// Monetary expansion formula parameter
+    pub monetary_expansion_decay: u64,
```
Comment (tomtau) [inline docs/modules/rewarding.md]: required ranges?

Comment (tomtau) [reply docs/modules/rewarding.md]: should `monetary_expansion_tau` be called `initial_tau` or something like that if it decays?

File: docs/modules/rewarding.md [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+    /// Reward pool other than the monetary expansion accumulated in current period
+    pub period_bonus: Coin,
+    /// last block height that updated it (i64 from Tendermint protobuf)
+    pub last_block_height: BlockHeight,
+    /// last reward distribution time, default to the genesis time
+    pub last_distribution_time: Timespec,
+    /// Record how many coins have been minted, can't exceed the cap
+    pub minted: Coin,
+    /// Parameter in monetary expansion formula, decayed for each rewards distribution
+    pub tau: u64,
```
Comment (tomtau) [inline docs/modules/rewarding.md]: `current_tau` to make it clearer that it may be different from the initial one?

File: docs/modules/rewarding.md [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+### ABCI Events
+
+#### Begin block
+
+```rust
+    fn begin_block(&mut self, req: &RequestBeginBlock) -> ResponseBeginBlock {
+        self.last_state
+            .as_mut()
+            .unwrap()
+            .validators
```
Comment (tomtau) [inline docs/modules/rewarding.md]: this is a sort of "implementation detail" and likely to change

File: docs/modules/rewarding.md [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    pub fn record_voters_for_rewarding(
+        &mut self,
+        addrs: impl Iterator<Item = TendermintValidatorAddress>,
+    ) {
+        for addr in addrs {
+            if let Some(staking_address) = self.tendermint_validator_addresses.get(&addr) {
+                if let Some(voting_power) = self
+                    .validator_state_helper
+                    .get_voting_power(staking_address)
+                {
```
Comment (tomtau) [inline docs/modules/rewarding.md]: same

**Review by tomtau — CHANGES_REQUESTED** [review]
the doc should be clear and compact, so one can easily judge whether the code does what it should do -- if there's a copy-pasted super-detailed code (rather than just high level snippets), it'll likely get out of date... and it may hard to judge whether the evolved code still does what it should be doing

File: docs/modules/rewarding.md (hunk 6 above)
Comment (tomtau) [reply docs/modules/rewarding.md]: many unnecessary details all below (plus the semantics of many of these detailed items isn't define, unless one looks at the code right now, so can be out of date -- e.g. `self.tendermint_validator_addresses`)

File: docs/modules/rewarding.md [hunk 7]
```diff
@@ -0,0 +1,304 @@
+# Rewarding
+
+## Abstract
+
+Rewarding module enable crypto chain to reward protocol participators, the design should meet following requirements:
+
+- Fixed total supply.
+- Fairness, all honest participators should get rewards, non-participators should not.
```
Comment (devashishdxt) [inline docs/modules/rewarding.md]: Define fairness?

File: docs/modules/rewarding.md [hunk 8]
```diff
@@ -0,0 +1,304 @@
+# Rewarding
+
+## Abstract
+
+Rewarding module enable crypto chain to reward protocol participators, the design should meet following requirements:
+
+- Fixed total supply.
```
Comment (devashishdxt) [inline docs/modules/rewarding.md]: Fixed total supply of what?

File: docs/modules/rewarding.md [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+- Fairness, all honest participators should get rewards, non-participators should not.
+- The rewards amount a validator received should be proportional to the voting power.
+
+The basic idea is to record signing voters for each block, along with the voting power at the vote time. Distribute the reward pool according to the voting powers.
```
Comment (devashishdxt) [inline docs/modules/rewarding.md]: Describe more with an example?

File: docs/modules/rewarding.md [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+#### Fixed point arithemetic
+
+TODO
+
+## Implementation
+
+### Network parameters
+
+```rust
+pub struct RewardsParameters {
```
Comment (devashishdxt) [inline docs/modules/rewarding.md:134]: At the start of the document, can you explain the role each of these network parameters and how they change reward behavior?

File: docs/modules/rewarding.md [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+
+The basic idea is to record signing voters for each block, along with the voting power at the vote time. Distribute the reward pool according to the voting powers.
+
+### Concepts
+
+#### Signing voters
+
+Signing voters are defined as:
+
+```rust
```
Comment (devashishdxt) [inline docs/modules/rewarding.md:44]: One needs to know **Rust** and **ABCI internals** to understand what are **Signing voters**. Maybe add a bit more description?

File: docs/modules/rewarding.md [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+  .last_commit_info
+  .votes
+  .filter(|vote| vote.signed_last_block)
+```
+
+All the signing voters of each block get rewarded, if a validator don't participate in voting, it doesn't get rewarded.
+
+#### Reward period
+
+Rewards are distributed to the validators periodically.
```
Comment (devashishdxt) [inline docs/modules/rewarding.md]: Also specify that this can be changed using network parameter in `genesis.json`.

File: docs/modules/rewarding.md [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+
+#### Reward pool
+
+During the reward period, the following sources of rewarding are accumulated into reward pool:
+
+- Monetary expansion
```
Comment (devashishdxt) [inline docs/modules/rewarding.md]: What does monetary expansion mean? Add a small definition and also link to the **Monetary expansion** section.

File: docs/modules/rewarding.md [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+}
+
+pub struct ChainNodeApp {
+    pub rewards_pool_updated: bool
+}
+```
+
+### ABCI Events
+
+#### Begin block
```
Comment (devashishdxt) [inline docs/modules/rewarding.md:221]: Maybe specify **_in English_**, what should happen during `BeginBlock` and don't expose the whole implementation (as it may change in future).

File: docs/modules/rewarding.md [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+```rust
+    pub fn commit_handler(&mut self, _req: &RequestCommit) -> ResponseCommit {
+        if self.rewards_pool_updated {
+            top_level.rewards_pool.last_block_height = new_state.last_block_height;
+            self.rewards_pool_updated = false;
+        }
+    }
+```
+
+### Hooks
```
Comment (devashishdxt) [inline docs/modules/rewarding.md:362]: No explanation of these **Hooks** and also no mention of where these should be called from.

File: docs/modules/rewarding.md [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+
+New coins are minted into the reward pool at the end of reward period. The amount is calculated based on the total staking of the last block:
+
+```
+R0 = rewards_config["monetary_expansion_r0"]
+tau = rewards_config["monetary_expansion_tau"]
+P = rewards_config["distribution_period"]
+
+# total bonded amount of the active validators at the end of the reward epoch
+S = total_staking
```
Comment (devashishdxt) [inline docs/modules/rewarding.md]: How is `total_staking` calculated?

**Review by devashishdxt — CHANGES_REQUESTED** [review]
There are still a few missing pieces which this spec does not explain:
- What happens when a validator gets jailed in a reward period?
- What happens when a validator unbonds their funds in a reward period?

Also, it is not clear what should happen in each ABCI call `InitChain`, `BeginBlock`, `EndBlock`, etc. One has to read and understand the whole code to know the behavior of rewards module.

I think it is better to merge this with [Validator Rewards](https://github.com/crypto-com/chain-docs/blob/master/docs/getting-started/staking.md#validator-rewards) section in main documentation (move everything to a new file?) which already explains everything in good detail.

File: docs/modules/rewarding.md (hunk 1 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: I changed to "The rewards amount distributed to a validator should be proportional to his voting power", the minted amount will be explained in the monetary expansion section, it's related to total staking, but not necessarily proportional.

File: docs/modules/rewarding.md (hunk 3 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: > required ranges?

Added the validation code.

Comment (yihuang) [reply docs/modules/rewarding.md]: > should `monetary_expansion_tau` be called `initial_tau` or something like that if it decays?

Since they are all in `InitNetworkParameters`, I think the init part if implicit already?

File: docs/modules/rewarding.md (hunk 4 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: Done

File: docs/modules/rewarding.md (hunk 7 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: The sentence after it explains it: "all honest participators should get rewards, non-participators should not".

File: docs/modules/rewarding.md (hunk 8 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: Changed it to "The total supply of CRO is fixed"

File: docs/modules/rewarding.md (hunk 11 above)
Comment (yihuang) [reply docs/modules/rewarding.md:44]: "All the active validators who follow the protocol honestly will be recorded as signing voters."
Added this sentence, should be enough for people to understand it.

File: docs/modules/rewarding.md (hunk 12 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: Done

File: docs/modules/rewarding.md (hunk 13 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: Added a link.

**Comment by yihuang** [comment]
Just cleaned up the code, removed the accidental things into separated functions.
Now there's very little thing that can be removed without affecting the logic.

File: docs/modules/rewarding.md (hunk 16 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: There is a "total staking" section now which explains it.

File: docs/modules/rewarding.md (hunk 15 above)
Comment (yihuang) [reply docs/modules/rewarding.md:362]: Added more descriptions in the hook section, please review to see if it's clear enough.

File: docs/modules/rewarding.md (hunk 14 above)
Comment (yihuang) [reply docs/modules/rewarding.md:221]: Added some simple descriptions before that code.

File: docs/modules/rewarding.md (hunk 10 above)
Comment (yihuang) [reply docs/modules/rewarding.md:134]: I think the parameters are all explained in different places in part 1.

**Comment by yihuang** [comment]
> There are still a few missing pieces which this spec does not explain:
> 
> * What happens when a validator gets jailed in a reward period?
> * What happens when a validator unbonds their funds in a reward period?

They are explained in the section hooks, added more descriptions.

> 
> Also, it is not clear what should happen in each ABCI call `InitChain`, `BeginBlock`, `EndBlock`, etc. One has to read and understand the whole code to know the behavior of rewards module.
> 

The code is cleaned up now, with comments, it should be quite readable to a developer.

> I think it is better to merge this with [Validator Rewards](https://github.com/crypto-com/chain-docs/blob/master/docs/getting-started/staking.md#validator-rewards) section in main documentation (move everything to a new file?) which already explains everything in good detail.

File: docs/modules/rewarding.md (hunk 5 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: Cleaned up the code, moved the accidental things into separated functions.

File: docs/modules/rewarding.md (hunk 6 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: Cleaned up the code, moved the accidental things into separated functions.

**Comment by yihuang** [comment]
> the doc should be clear and compact, so one can easily judge whether the code does what it should do -- if there's a copy-pasted super-detailed code (rather than just high level snippets), it'll likely get out of date... and it may hard to judge whether the evolved code still does what it should be doing

I think with the cleaned-up version of code, there is very little that is not directly related to the logic. And it's the more precise way to describe the logic.

File: docs/modules/rewarding.md (hunk 7 above)
Comment (tomtau) [reply docs/modules/rewarding.md]: where does this definition come from? 
the common one is:
```
3 (fairness). A process gets a reward for a block if and only if it has a reward parameter for that block equal to 1. 
3' (eventual fairness). There exists a height H such that for a block in the blockchain at height H′ > H a process gets a reward for that block if and only if it has a reward parameter for that block equal to 1.
```

File: docs/modules/rewarding.md [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+
+            for (addr, &power) in self.validator_state().signed_voters.iter() {
+                // Use u128 to prevent overflow of intermidiate results
+                let amount: u128 =
+                    u128::from(reward_pool) * u128::from(power) / u128::from(sum_power);
+                let amount = Coin::new(amount.try_into().unwrap()).unwrap();
+                remains = remains.saturating_sub(amount);
+                distribution.push((*addr, amount));
+            }
+            remains
```
Comment (tomtau) [inline docs/modules/rewarding.md]: this is still "unfair" (as per the agreed definition):

1. it only tracks "IF" the validator correctly participated, not how often
2. it distributes per vote power share, not per participation.

--
(eventually) fair reward distribution should reward a node for each block if and only if it participated correctly (live, non-byzantine).

...
Here, the counter-example is that if there are more than 1 block during that period, it won't be fair. e.g. one node behaves like this:
block 1 -- reward parameter=1
block 2 -- reward parameter=0
block 3 -- reward parameter=0
...
blocks N -> distribution -- gets rewarded for all blocks up to N, regardless how it behaved in each of them which is unfair.

File: docs/modules/rewarding.md [hunk 18]
```diff
@@ -0,0 +1,334 @@
+# Rewarding
+
+## Abstract
+
+Rewarding module enable crypto chain to reward protocol participators, the design should meet following requirements:
+
+- The total supply of CRO is fixed.
+- Fairness, all honest participators should get rewards, non-participators should not.
+- The rewards amount distributed to a validator should be proportional to his voting power.
```
Comment (tomtau) [inline docs/modules/rewarding.md]: The total reward emitted should be proportional to the total voting power,

but the distributed amount should be proportional to the correct participation (assuming distribution doesn't happen every block due to performance degrade).

**Review by tomtau — CHANGES_REQUESTED** [review]
looks a bit simpler, but again the code snippets should ideally omit details that may change or explain them more in detail.
e.g. `self.rewards_pool_updated` is just some temporary flag (the implementation may change) that helps with app hash re-calculations or what not?

File: docs/modules/rewarding.md (hunk 18 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: total reward emitted could be positively or negatively related to the total voting power, not necessarily proportional(maybe I have a wrong understanding of proportional?), also it could be unrelated to total staking at all if monetary expansion cap is reached.
"correct participation" is the second point?

**Comment by yihuang** [comment]
> looks a bit simpler, but again the code snippets should ideally omit details that may change or explain them more in detail.
> e.g. `self.rewards_pool_updated` is just some temporary flag (the implementation may change) that helps with app hash re-calculations or what not?

`self.rewards_pool_updated` is part of the logic of how `last_block_height` gets updated, I don't see a way to explain it without this flag.

File: docs/modules/rewarding.md (hunk 17 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: You seem to try to distribute rewards per block, but I think they are actually equivalent:
```
block 1 -- reward parameter=1, power of A=p, power of others=p'
block 2 -- reward parameter=0
block 3 -- reward parameter=0
...
blocks N -> distribution
``` 
Assuming all the other validators have reward parameter=1, then the reward distribution of A is:

```
total_power = p + p' * N

# current computation
reward_a = reward_pool * p / total_power

# only focus on block 1
reward_block1 = reward_pool * (p + p') / total_power
reward_a = reward_block1 * p / (p + p')
         = reward_pool * p / total_power
```

Comment (yihuang) [reply docs/modules/rewarding.md]: > where does this definition come from?
> the common one is:
> 
> 3 (fairness). A process gets a reward for a block if and only if it has a reward parameter for that block equal to 1. 
> 3' (eventual fairness). There exists a height H such that for a block in the blockchain at height H′ > H a process gets a reward for that block if and only if it has a reward parameter for that block equal to 1.

I think it's just the 3 in layman's term?

File: docs/modules/rewarding.md (hunk 10 above)
Comment (devashishdxt) [reply docs/modules/rewarding.md:134]: Where is part 1? Am I missing something?

Comment (yihuang) [reply docs/modules/rewarding.md:134]: > Where is part 1? Am I missing something?

I mean the section "Abstract".
For example, the monetary expansion parameters R0/tau are in the monetary expansion formula.
The reward period has a section in concepts. I should add more description for the `rewards_monetary_expansion_cap` and `monetary_expansion_decay `.

File: docs/modules/rewarding.md (hunk 9 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: Added an example now.

File: docs/modules/rewarding.md (hunk 10 above)
Comment (yihuang) [reply docs/modules/rewarding.md:134]: Done

File: docs/modules/rewarding.md (hunk 17 above)
Comment (tomtau) [reply docs/modules/rewarding.md]: normal fairness gets that reward for that block.

eventual fairness gets reward for that block in some time in the future -- the important bit here there is " if and only if it ".

In the current mechanism, yes, it gets a reward for that block AND it also gets rewards for blocks where its reward parameter=0 (as the current mechanism doesn't distinguish correct participation) which it shouldn't

File: docs/modules/rewarding.md (hunk 18 above)
Comment (tomtau) [reply docs/modules/rewarding.md]: I'm confused if the third point talks about the share or the total emitted reward?

File: docs/modules/rewarding.md (hunk 17 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: > normal fairness gets that reward for that block.
> 
> eventual fairness gets reward for that block in some time in the future -- the important bit here there is " if and only if it ".
> 
> In the current mechanism, yes, it gets a reward for that block AND it also gets rewards for blocks where its reward parameter=0 (as the current mechanism doesn't distinguish correct participation) which it shouldn't

As I showed in the previous comment, it only gets the reward for blocks where its reward parameter = 1, not get rewards for blocks where its reward parameter=0.
I copy it here:

> You seem to try to distribute rewards per block, but I think they are actually equivalent:
> ```
> block 1 -- reward parameter=1, power of A=p, power of others=p'
> block 2 -- reward parameter=0
> block 3 -- reward parameter=0
> ...
> blocks N -> distribution
> ``` 
> Assuming all the other validators have reward parameter=1, then the reward distribution of A is:
> 
> ```
> total_power = p + p' * N
> 
> # current computation
> reward_a = reward_pool * p / total_power
> 
> # only focus on block 1
> reward_block1 = reward_pool * (p + p') / total_power
> reward_a = reward_block1 * p / (p + p')
>          = reward_pool * p / total_power
> ```
> 
>

File: docs/modules/rewarding.md (hunk 18 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: > I'm confused if the third point talks about the share or the total emitted reward?

It says "The rewards amount distributed to a validator", so it's the share of a single validator, how should I make it clearer?

File: docs/modules/rewarding.md (hunk 17 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: Do you mean don't accumulate rewards into a pool for the whole period, but calculate reward distribution for each block separately. Because different blocks might have different fees collected?

File: docs/modules/rewarding.md (hunk 18 above)
Comment (tomtau) [reply docs/modules/rewarding.md]: Well, one of the points of this was:

1. total emitted amount should be proportional to the total vote power (as before)
2. individual shares should correspond to correct behaviour participation, *regardless* of the vote power

And the correct behaviour is defined with the respect to individual blocks with "IF AND ONLY IF" for each block -- it's not defined as "As long as a validator exhibited correct behaviour in 1 block out of N in a given, he should get reward for all N blocks"

File: docs/modules/rewarding.md (hunk 17 above)
Comment (tomtau) [reply docs/modules/rewarding.md]: it's defined with "IF AND ONLY IF" -- e.g. two validators A B, both have the same vote power.
Node A:
```
block 1 -- reward parameter=1
block 2 -- reward parameter=0
block 3 -- reward parameter=0
...
```
Node B:
```
block 1 -- reward parameter=1
block 2 -- reward parameter=1
block 3 -- reward parameter=1
...
```

reward gets distributed at N -- currently, the share calculation takes into account only vote power... so here, A and B have the same vote power, so they get rewarded the same.
When interpreted per block, this does not satisfy (eventual) fairness, as participants should be reward IF AND ONLY IF... here, A was rewarded for blocks 2, 3...

File: docs/modules/rewarding.md (hunk 18 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: “total emitted amount should be proportional to the total vote power” is not accurate, fees and slashings are not related to total vote power, the amount of new minted coins is a function of total vote power, but not necessarily proportional to that.

By "regardless of the vote power", do you mean one vote of M power and one vote of N power should get the same amount of reward?

"he should get reward for all N blocks", I don't understand, in current computation, he indeed only get the rewards for the blocks he participated correctly.

File: docs/modules/rewarding.md (hunk 17 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: > it's defined with "IF AND ONLY IF" -- e.g. two validators A B, both have the same vote power.
> Node A:
> 
> ```
> block 1 -- reward parameter=1
> block 2 -- reward parameter=0
> block 3 -- reward parameter=0
> ...
> ```
> 
> Node B:
> 
> ```
> block 1 -- reward parameter=1
> block 2 -- reward parameter=1
> block 3 -- reward parameter=1
> ...
> ```
> 
> reward gets distributed at N -- currently, the share calculation takes into account only vote power... so here, A and B have the same vote power, so they get rewarded the same.
> When interpreted per block, this does not satisfy (eventual) fairness, as participants should be reward IF AND ONLY IF... here, A was rewarded for blocks 2, 3...

In this case, the computation is like this:
```
pa = power of A
pb = power of B
sum_power = pa + pb * N
sum_power_a = pa
sum_power_b = pb * N
reward_a = reward_pool * sum_power_a / sum_power
reward_b = reward_pool * sum_power_b / sum_power
```

if interpret it per block:
```
power_block1 = pa + pb
sum_power = pa + pb * N
reward_block1 = reward_pool * power_block1 / sum_power
                = reward_pool * (pa + pb) / sum_power
reward_a = reward_block1 * pa / (pa + pb)
           = reward_pool * pa / sum_power
```

The result is the same.

File: docs/modules/rewarding.md (hunk 18 above)
Comment (tomtau) [reply docs/modules/rewarding.md]: > “total emitted amount should be proportional to the total vote power” is not accurate, fees and slashings are not related to total vote power, the amount of new minted coins is a function of total vote power, but not necessarily proportional to that.

emitted means it's newly emitted, so doesn't include fees or slashing -- perhaps more precise is inversely proportional or what's out of the formula.
it can say that the "total reward includes total emitted coins (proportional to ...) and bonus"

> By "regardless of the vote power", do you mean one vote of M power and one vote of N power should get the same amount of reward?

It depends on their correct behaviour.
If both behaved correctly (live, non-byzantine) at all times, yes

File: docs/modules/rewarding.md (hunk 17 above)
Comment (tomtau) [reply docs/modules/rewarding.md]: Oh, https://github.com/crypto-com/chain-docs/pull/94/files#diff-2c819a76ed9cc68e5c8a8d51dc8ac205R208 there's a `saturating_add`

Comment (tomtau) [reply docs/modules/rewarding.md]: I skimmed through it and only saw "insert"

File: docs/modules/rewarding.md (hunk 18 above)
Comment (yihuang) [reply docs/modules/rewarding.md]: > > By "regardless of the vote power", do you mean one vote of M power and one vote of N power should get the same amount of reward?
> 
> It depends on their correct behaviour.
> If both behaved correctly (live, non-byzantine) at all times, yes

Do you mean a validator with 10 power get the same amount of rewards as a validator with 1 power, if they both behaved correctly at all times? Isn't this incentivizes the validator with more power to separate his power into multiple smaller validators, and takes more validator spots, and which is something we want to avoid?

File: docs/modules/rewarding.md (hunk 17 above)
Comment (tomtau) [reply docs/modules/rewarding.md]: Anyway, this shows that a more high-level concise description or pseudocode will be clearer than 100+ lines of detailed code where it is hard to spot the important parts

Comment (tomtau) [reply docs/modules/rewarding.md]: `record_voters_for_rewarding` could just be something like:
```
if signed_voters[staked_address]:
signed_voters[staked_address] += vote_power_at_X
else:
signed_voters[staked_address] = vote_power_at_X
```

Comment (yihuang) [reply docs/modules/rewarding.md]: > `record_voters_for_rewarding` could just be something like:
> 
> ```
> if signed_voters[staked_address]:
> signed_voters[staked_address] += vote_power_at_X
> else:
> signed_voters[staked_address] = vote_power_at_X
> ```

But it will miss two informations:
- overflow behavior
- Error handling when address and voting power not found

**Review by devashishdxt — COMMENTED** [review]
A few thoughts:

This document exposes a lot of implementation details (it exposes the whole implementation). Specifications, generally, are supposed to act as a blueprint to help developers implement the functionality. It should describe all the flows in the functionality and also specify behavior of the program for each flow. Besides this, it is ok to provide high level interfaces that can be used by developer but, detailed implementation shouldn't be a part of specification.

Understanding the behaviour of rewards module from this document requires going through the code and understanding it, which is not what specification is meant for. Ideally, `specification + developer = code`. Whereas here, `specification + code + developer = more code`.

Any thoughts?

**Comment by yihuang** [comment]
> A few thoughts:
> 
> This document exposes a lot of implementation details (it exposes the whole implementation). Specifications, generally, are supposed to act as a blueprint to help developers implement the functionality. It should describe all the flows in the functionality and also specify behavior of the program for each flow. Besides this, it is ok to provide high level interfaces that can be used by developer but, detailed implementation shouldn't be a part of specification.
> 
> Understanding the behaviour of rewards module from this document requires going through the code and understanding it, which is not what specification is meant for. Ideally, `specification + developer = code`. Whereas here, `specification + code + developer = more code`.
> 
> Any thoughts?

I guess it all come down to how precise do we want the specification to be, or what functionality do we want from the specification, do we want it to be some descriptive documentation, or precise stuff which can be used to check the correctness.
For example, the error handling(panic or not in some edge cases) or overflow behavior(panic, wrap, saturate) is important part of the correctness, many seemingly verboseness in the code are  actually these things.

**Comment by devashishdxt** [comment]
> > A few thoughts:
> > This document exposes a lot of implementation details (it exposes the whole implementation). Specifications, generally, are supposed to act as a blueprint to help developers implement the functionality. It should describe all the flows in the functionality and also specify behavior of the program for each flow. Besides this, it is ok to provide high level interfaces that can be used by developer but, detailed implementation shouldn't be a part of specification.
> > Understanding the behaviour of rewards module from this document requires going through the code and understanding it, which is not what specification is meant for. Ideally, `specification + developer = code`. Whereas here, `specification + code + developer = more code`.
> > Any thoughts?
> 
> I guess it all come down to how precise do we want the specification to be, or what functionality do we want from the specification, do we want it to be some descriptive documentation, or precise stuff which can be used to check the correctness.
> For example, the error handling(panic or not in some edge cases) or overflow behavior(panic, wrap, saturate) is important part of the correctness, many seemingly verboseness in the code are actually these things.

These are important parts of a program's implementation and most of the specifications don't delve into these implementation details because the idea to to keep them agnostic of how any developer chooses to implement them (for example, decisions like which data sctucture to use, what language to use, how to handle errors, etc. are left to developers). Here is an example, [this](https://www.ietf.org/rfc/rfc793.txt) is **Transmission Control Protocol** specification and they explain the whole protocol without code.

P.S.: I'm ok with writing the whole code in specification if everyone is comfortable with this.

File: docs/modules/rewarding.md (hunk 18 above)
Comment (tomtau) [reply docs/modules/rewarding.md]: Ignore that -- it was my misunderstanding of how the activity is recorded.

File: docs/modules/rewarding.md (hunk 17 above)
Comment (tomtau) [reply docs/modules/rewarding.md]: Is any of them expected to happen as a part of normal behaviour?

It'll ideally list all its assumptions / invariants separately -- e.g. it's assumed that one can lookup staking address (based on the consensus pubkey/address) for the duration of the reward period, unless it's been removed due to punishment mechanisms?

For expected behaviour, it should specify these.
For unexpected behaviour, there should be an overall note around handling -- e.g. unexpected behaviour / invariant breaking behaviour will be logged with "error" level in deployment builds, but will panic in debug/fuzzing/test builds?

**Review by tomtau — CHANGES_REQUESTED** [review]
the revisions look better -- it'll be more comprehensive if the code snippets are shorter and only capture the important part.
also some details may be missing ? e.g. `add_rewards` isn't documented -- nonce is increased by 1? something else?

Everything that's not common Rust, Tendermint ABCI or defined elsewhere in chain-docs is subject to change (all these `self.<something>` ...), so it's better to be omitted and rephrased in more future-proof way -- a sort of: here are the inputs, here are assumptions / invariants about these inputs, this is how they are processed, these are the outputs.

File: docs/modules/rewarding.md [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+
+#### Reward pool
+
+During the reward period, the following sources of funds are accumulated into reward pool:
+
+- [Monetary expansion](#monetary-expansion)
+- [Transaction Fees](../getting-started/transaction.md#transaction-fee)
+- Slashings
+
+The reward pool get accumulated through the whole reward period, it get distributed once at the end of reward period.
```
Comment (tomtau) [inline docs/modules/rewarding.md]:
```suggestion
The reward pool gets accumulated throughout the whole reward period; it gets distributed once at the end of each reward period.
```

File: docs/modules/rewarding.md [hunk 20]
```diff
@@ -0,0 +1,389 @@
...
+
+Because of integer calculations, there will be a few remained funds after distribution, they should be left to the next period.
```
Comment (tomtau) [inline docs/modules/rewarding.md]:
```suggestion
Due to fixed point calculations, there may be a few remainders from arithmetic operations after distribution. These remainder amounts should be left in the rewards pool for the next period.
```

File: docs/modules/rewarding.md [hunk 21]
```diff
@@ -0,0 +1,389 @@
...
+
+If there is zero participator for the period, the rewards pool should be left to the next period, rather than get burned.
```
Comment (tomtau) [inline docs/modules/rewarding.md:67]: does this mean that if there were no participants, reward distribution should not happen? When would this happen?

File: docs/modules/rewarding.md [hunk 22]
```diff
@@ -0,0 +1,389 @@
...
+The main actions all happens in begin block event, it's the time we get reports of participators through `last_commit_info`, also the time the node state is consistent.
+
+1. Collect the validator addresses of reward participators from `last_commit_info`.
+2. Record them, and accumulate their voting powers.
+3. If a reward period has passed, distribute the reward pool according to the participator statistics.
+
+##### `begin_block`
+
+```rust
+fn begin_block(req: RequestBeginBlock) {
```
Comment (tomtau) [inline docs/modules/rewarding.md:232]: just side-note that I guess it should happen after punishment-related begin block processing? there's a separate issue for this / how to generate validator updates: https://github.com/crypto-com/chain-docs/issues/81 so no need in this PR

**Review by tomtau — CHANGES_REQUESTED** [review]
looks all right and much clearer -- left a few small suggestions, but otherwise good to go

File: docs/modules/rewarding.md (hunk 21 above)
Comment (yihuang) [reply docs/modules/rewarding.md:67]: Maybe, all the previous participators get jailed in node state somehow, but still manage to produce the last block of reward period, because tendermint validator set don't updated yet.
